### ISSUE TITLE
Couldn't Process XML into a Doc, Premature End of File, and HTTP 401 Response Code

### ISSUE THREAD

**Issue by Kashinoga** [googleads/googleads-java-lib#138]
Hello,

When attempting this call:

> 
>     ReportDownloadResponse response =
>         new ReportDownloader(session).downloadReport(reportDefinition);

We are met with this error log:

> [Fatal Error] :1:1: Premature end of file.
> 2018-04-20 08:09:06 [main] ERROR c.g.a.a.a.l.c.A.soapXmlLogger.extract(85) - Couldn't process XML into a Document
> org.xml.sax.SAXParseException: Premature end of file.
> 	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(Unknown Source)
> 	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(Unknown Source)
> 	at javax.xml.parsers.DocumentBuilder.parse(Unknown Source)
> 	at com.google.api.ads.common.lib.utils.XmlFieldExtractor.extract(XmlFieldExtractor.java:72)
> 	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelperImpl.handleResponse(AdHocReportDownloadHelperImpl.java:119)
> 	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelperImpl.downloadReport(AdHocReportDownloadHelperImpl.java:60)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
> 	at java.lang.reflect.Method.invoke(Unknown Source)
> 	at com.google.api.ads.common.lib.utils.AdsUtilityInvocationHandler.handleInvocation(AdsUtilityInvocationHandler.java:46)
> 	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:84)
> 	at com.sun.proxy.$Proxy46.downloadReport(Unknown Source)
> 	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelper.downloadReport(AdHocReportDownloadHelper.java:46)
> 	at com.google.api.ads.adwords.lib.utils.v201710.ReportDownloader.downloadReport(ReportDownloader.java:65)
> 	at com.zirous.indago.google.adwords.AdwordsTask.runAdTask(AdwordsTask.java:103)
> 	at com.zirous.indago.google.adwords.AdwordsTask.run(AdwordsTask.java:63)
> 	at com.zirous.indago.google.adwords.AdwordsTests.main(AdwordsTests.java:27)
> HTTP Response Code: 401
> 	at com.google.api.ads.adwords.lib.utils.v201710.DetailedReportDownloadResponseException$Builder.build(DetailedReportDownloadResponseException.java:35)
> 	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelperImpl.handleResponse(AdHocReportDownloadHelperImpl.java:117)
> 	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelperImpl.downloadReport(AdHocReportDownloadHelperImpl.java:60)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
> 	at java.lang.reflect.Method.invoke(Unknown Source)
> 	at com.google.api.ads.common.lib.utils.AdsUtilityInvocationHandler.handleInvocation(AdsUtilityInvocationHandler.java:46)
> 	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:84)
> 	at com.sun.proxy.$Proxy46.downloadReport(Unknown Source)
> 	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelper.downloadReport(AdHocReportDownloadHelper.java:46)
> 	at com.google.api.ads.adwords.lib.utils.v201710.ReportDownloader.downloadReport(ReportDownloader.java:65)
> 	at com.zirous.indago.google.adwords.AdwordsTask.runAdTask(AdwordsTask.java:103)
> 	at com.zirous.indago.google.adwords.AdwordsTask.run(AdwordsTask.java:63)
> 	at com.zirous.indago.google.adwords.AdwordsTests.main(AdwordsTests.java:27)
> Exception in thread "main" java.lang.RuntimeException: 
> 	at com.zirous.indago.google.adwords.AdwordsTask.run(AdwordsTask.java:66)
> 	at com.zirous.indago.google.adwords.AdwordsTests.main(AdwordsTests.java:27)

We ran a couple of test calls in Postman to download a report with a basic report definition and was met with a 502 Bad Gateway error. The Adwords forum suggested that it might be a client library issue instead.

Our reports were downloading fine about a week ago, then these errors happened without any change on our end.

Any ideas on what might be the cause? 

Thank you.

**Comment by jradcliff** [comment]
Hi,

One suggestion is to enable `FINE` debugging on the logger named `report_download` logger. For example, if you're using the [`log4j.properties` included with the examples](https://github.com/googleads/googleads-java-lib/blob/master/examples/adwords_axis/src/main/resources/log4j.properties), you can replace `INFO` with `FINE` on the following two lines:

Logger config:

    log4j.category.com.google.api.ads.adwords.lib.utils.report_download=INFO, CONSOLE

Console config:

    log4j.appender.CONSOLE.Threshold=INFO

If you share the resulting output here, please make sure to remove any sensitive information such as your developer token.

Could you share the link from the AdWords API forum discussion as well?

Thanks,
Josh, AdWords API Team

**Comment by Kashinoga** [comment]
Hello,

This is the forum discussion: https://groups.google.com/forum/#!topic/adwords-api/jpq24r06b2M

This was a private response (no sensitive info) from them after providing some logs: 


> It seems like the error you are encountering is more related to the Java client library rather than the AdWords API. I would suggest to post your query here as the client library owners are better equipped to assist you in this matter.

Here is the log after adding a FINE level to a logger "com.google.api" since we're currently using java.util.logging.Logger:

> 2018-04-23 11:32:49 [main] DEBUG c.g.a.a.c.l.u.l.A.ads_api_lib.createCombinedConfiguration(158) - Could not load optional configuration: ConfigurationInfo{location=ads.properties, isOptional=true}
2018-04-23 11:32:49 [main] DEBUG c.g.a.a.c.l.u.l.A.ads_api_lib.createCombinedConfiguration(158) - Could not load optional configuration: ConfigurationInfo{location=ads.properties, isOptional=true}
2018-04-23 11:32:49 [main] DEBUG c.g.a.a.c.l.u.l.A.ads_api_lib.createCombinedConfiguration(158) - Could not load optional configuration: ConfigurationInfo{location=ads.properties, isOptional=true}
2018-04-23 11:32:50 [main] WARN  c.g.a.a.a.lib.utils.report_download.logRequestSummary(148) - Request made: Service: reportdownload Method: POST clientCustomerId: REDACTED URL: https://adwords.google.com//api/adwords/reportdownload/v201710 Request ID: null ResponseTime(ms): null OperationsCount: null IsFault: true FaultMessage: com.google.api.ads.adwords.lib.utils.ReportException: 401: Unauthorized
2018-04-23 11:32:50 [main] INFO  c.g.a.a.a.lib.utils.report_download.logRequestDetails(114) - HTTP request:
accept-encoding: [gzip]
authorization: REDACTED
user-agent: [unknown (AwApi-Java, Common-Java/3.10.0, Java/1.8.0_111, maven, ReportDownloader)]
developertoken: REDACTED
clientcustomerid: REDACTED
skipreportheader: false
skipcolumnheader: false
skipreportsummary: false
includezeroimpressions: false

> Content:
__rdxml: <?xml version="1.0" encoding="UTF-8" standalone="yes"?><reportDefinition xmlns:ns2="https://adwords.google.com/api/adwords/cm/v201710"><ns2:selector><ns2:fields>Date</ns2:fields><ns2:fields>Id</ns2:fields><ns2:fields>AdNetworkType1</ns2:fields><ns2:fields>AdNetworkType2</ns2:fields><ns2:fields>CampaignId</ns2:fields><ns2:fields>CampaignName</ns2:fields><ns2:fields>AdGroupId</ns2:fields><ns2:fields>AdGroupName</ns2:fields><ns2:fields>AdType</ns2:fields><ns2:fields>CreativeFinalUrls</ns2:fields><ns2:fields>Description</ns2:fields><ns2:fields>Interactions</ns2:fields><ns2:fields>Clicks</ns2:fields><ns2:fields>Impressions</ns2:fields><ns2:fields>Cost</ns2:fields><ns2:fields>CreativeUrlCustomParameters</ns2:fields><ns2:fields>CriterionId</ns2:fields><ns2:fields>CriterionType</ns2:fields><ns2:dateRange><ns2:min>20180422</ns2:min><ns2:max>20180422</ns2:max></ns2:dateRange></ns2:selector><ns2:reportName>AD_PERFORMANCE_REPORT</ns2:reportName><ns2:reportType>AD_PERFORMANCE_REPORT</ns2:reportType><ns2:dateRangeType>CUSTOM_DATE</ns2:dateRangeType><ns2:downloadFormat>XML</ns2:downloadFormat></reportDefinition>

> 2018-04-23 11:32:50 [main] INFO  c.g.a.a.a.lib.utils.report_download.logRequestDetails(115) - HTTP response:
401 Unauthorized
cache-control: [private, max-age=0]
content-encoding: [gzip]
content-type: [text/html; charset=UTF-8]
date: [Mon, 23 Apr 2018 16:31:22 GMT]
expires: [Mon, 23 Apr 2018 16:31:22 GMT]
www-authenticate: [OAuth realm="https://accounts.google.com/OAuthGetRequestToken"]
transfer-encoding: [chunked]
alt-svc: [hq=":443"; ma=2592000; quic=51303433; quic=51303432; quic=51303431; quic=51303339; quic=51303335,quic=":443"; ma=2592000; v="43,42,41,39,35"]
server: [GSE]
x-content-type-options: [nosniff]
x-frame-options: [SAMEORIGIN]
x-xss-protection: [1; mode=block]

> Content:
REDACTED REPORT DATA
[Fatal Error] :1:1: Premature end of file.
2018-04-23 11:32:50 [main] ERROR c.g.a.a.a.l.c.A.soapXmlLogger.extract(85) - Couldn't process XML into a Document
org.xml.sax.SAXParseException: Premature end of file.
	at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(Unknown Source)
	at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(Unknown Source)
	at javax.xml.parsers.DocumentBuilder.parse(Unknown Source)
	at com.google.api.ads.common.lib.utils.XmlFieldExtractor.extract(XmlFieldExtractor.java:72)
	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelperImpl.handleResponse(AdHocReportDownloadHelperImpl.java:119)
	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelperImpl.downloadReport(AdHocReportDownloadHelperImpl.java:60)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at com.google.api.ads.common.lib.utils.AdsUtilityInvocationHandler.handleInvocation(AdsUtilityInvocationHandler.java:46)
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:84)
	at com.sun.proxy.$Proxy46.downloadReport(Unknown Source)
	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelper.downloadReport(AdHocReportDownloadHelper.java:46)
	at com.google.api.ads.adwords.lib.utils.v201710.ReportDownloader.downloadReport(ReportDownloader.java:65)
	at com.zirous.indago.google.adwords.AdwordsTask.runAdTask(AdwordsTask.java:103)
	at com.zirous.indago.google.adwords.AdwordsTask.run(AdwordsTask.java:63)
	at com.zirous.indago.google.adwords.AdwordsTests.main(AdwordsTests.java:27)
HTTP Response Code: 401
	at com.google.api.ads.adwords.lib.utils.v201710.DetailedReportDownloadResponseException$Builder.build(DetailedReportDownloadResponseException.java:35)
	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelperImpl.handleResponse(AdHocReportDownloadHelperImpl.java:117)
	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelperImpl.downloadReport(AdHocReportDownloadHelperImpl.java:60)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at com.google.api.ads.common.lib.utils.AdsUtilityInvocationHandler.handleInvocation(AdsUtilityInvocationHandler.java:46)
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:84)
	at com.sun.proxy.$Proxy46.downloadReport(Unknown Source)
	at com.google.api.ads.adwords.lib.utils.AdHocReportDownloadHelper.downloadReport(AdHocReportDownloadHelper.java:46)
	at com.google.api.ads.adwords.lib.utils.v201710.ReportDownloader.downloadReport(ReportDownloader.java:65)
	at com.zirous.indago.google.adwords.AdwordsTask.runAdTask(AdwordsTask.java:103)
	at com.zirous.indago.google.adwords.AdwordsTask.run(AdwordsTask.java:63)
	at com.zirous.indago.google.adwords.AdwordsTests.main(AdwordsTests.java:27)
Exception in thread "main" java.lang.RuntimeException: 
	at com.zirous.indago.google.adwords.AdwordsTask.run(AdwordsTask.java:66)
	at com.zirous.indago.google.adwords.AdwordsTests.main(AdwordsTests.java:27)

**Comment by jradcliff** [comment]
Thanks for the additional info. I'm not familiar with Postman, but did you try using the Postman Console to investigate the 502 errors?

https://www.getpostman.com/docs/v6/postman/sending_api_requests/troubleshooting_api_requests

Also, looking at your log files, the `__rdxml` looks incorrect:

```
__rdxml: ns2:selectorns2:fieldsDate</ns2:fields>ns2:fieldsId</ns2:fields>ns2:fieldsAdNetworkType1</ns2:fields>ns2:fieldsAdNetworkType2</ns2:fields>ns2:fieldsCampaignId</ns2:fields>ns2:fieldsCampaignName</ns2:fields>ns2:
```

Notice that it starts with `ns2:selctorns2:fieldsDate<...`, which is malformed XML, as the error indicates. I would expect it to start with the XML declaration like this:

```
__rdxml: <?xml version="1.0" encoding="UTF-8" standalone="yes"?><reportDefinition xmlns:ns2="https://adwords.google.com/api/adwords/cm/v201802"><ns2:selector><ns2:fields>...
```

Are you following the [Thread Safety guidelines for the library](https://github.com/googleads/googleads-java-lib/wiki/Thread-Safety)? I ask because I've seen odd issues like this when users have not followed those guidelines and shared library objects that are not thread safe.

Thanks,
Josh, AdWords API Team

**Comment by Kashinoga** [comment]
No problem. I did try investigating the 502 Bad Gateway errors, but always ended up with either "Try again in 30 second" messages or that something was up on the Adwords side.

About the _rdxml, something happened when I pasted the log (my bad, used quote markup instead of code), it looks like this currently:

```
Content:
__rdxml: <?xml version="1.0" encoding="UTF-8" standalone="yes"?><reportDefinition xmlns:ns2="https://adwords.google.com/api/adwords/cm/v201710"><ns2:selector><ns2:fields>Date</ns2:fields><ns2:fields>Id</ns2:fields><ns2:fields>AdNetworkType1</ns2:fields><ns2:fields>AdNetworkType2</ns2:fields><ns2:fields>CampaignId</ns2:fields><ns2:fields>CampaignName</ns2:fields><ns2:fields>AdGroupId</ns2:fields><ns2:fields>AdGroupName</ns2:fields><ns2:fields>AdType</ns2:fields><ns2:fields>CreativeFinalUrls</ns2:fields><ns2:fields>Description</ns2:fields><ns2:fields>Interactions</ns2:fields><ns2:fields>Clicks</ns2:fields><ns2:fields>Impressions</ns2:fields><ns2:fields>Cost</ns2:fields><ns2:fields>CreativeUrlCustomParameters</ns2:fields><ns2:fields>CriterionId</ns2:fields><ns2:fields>CriterionType</ns2:fields><ns2:dateRange><ns2:min>20180422</ns2:min><ns2:max>20180422</ns2:max></ns2:dateRange></ns2:selector><ns2:reportName>AD_PERFORMANCE_REPORT</ns2:reportName><ns2:reportType>AD_PERFORMANCE_REPORT</ns2:reportType><ns2:dateRangeType>CUSTOM_DATE</ns2:dateRangeType><ns2:downloadFormat>XML</ns2:downloadFormat></reportDefinition>
```

I'll have a look at that thread safety guidelines and get back to you.

Thank you.

**Comment by Kashinoga** [comment]
The guidelines are being followed in the way that nothing is being run at the same time.

**Comment by jradcliff** [comment]
Thanks for double-checking that. I noticed the following while looking at your logs:

*  In the detailed log you sent, the request is being sent to the **v201710** endpoint:

        URL: https://adwords.google.com//api/adwords/reportdownload/v201710 

* The XML you provided is using the **v201802** namespace.

Are you perhaps using the wrong version of the [`ReportDownloader` utility](https://github.com/googleads/googleads-java-lib/blob/master/modules/ads_lib/src/main/java/com/google/api/ads/adwords/lib/utils/v201802/ReportDownloader.java), or the wrong version of [`ReportDefinition`](https://github.com/googleads/googleads-java-lib/blob/master/modules/ads_lib/src/main/java/com/google/api/ads/adwords/lib/jaxb/v201802/ReportDefinition.java)? The v201710 endpoint will only accept a v201710 `ReportDefinition`.

Thanks,
Josh, AdWords API Team

**Comment by Kashinoga** [comment]
That is odd. I did previously update the pom to try the latest version of this SDK, but that did not resolve the problem so I reverted back to 3.10.0.

These are our dependencies and import statements relating to Adwords with a versioning number in them:

```
    <dependency>
      <groupId>com.google.api-ads</groupId>
      <artifactId>ads-lib</artifactId>
      <version>3.10.0</version>
    </dependency>
    <dependency>
      <groupId>com.google.api-ads</groupId>
      <artifactId>adwords-axis</artifactId>
      <version>3.10.0</version>
    </dependency>
```

```
import com.google.api.ads.adwords.lib.jaxb.v201710.DateRange;
import com.google.api.ads.adwords.lib.jaxb.v201710.DownloadFormat;
import com.google.api.ads.adwords.lib.jaxb.v201710.ReportDefinition;
import com.google.api.ads.adwords.lib.jaxb.v201710.ReportDefinitionDateRangeType;
import com.google.api.ads.adwords.lib.jaxb.v201710.ReportDefinitionReportType;
import com.google.api.ads.adwords.lib.jaxb.v201710.Selector;
import com.google.api.ads.adwords.lib.utils.v201710.ReportDownloader;
```

I am not sure how these and these statements:

```
// Create report definition.
ReportDefinition reportDefinition = new ReportDefinition();
reportDefinition.setReportName("AD_PERFORMANCE_REPORT");
reportDefinition.setDateRangeType(ReportDefinitionDateRangeType.CUSTOM_DATE);
reportDefinition.setReportType(ReportDefinitionReportType.AD_PERFORMANCE_REPORT);
reportDefinition.setDownloadFormat(DownloadFormat.XML);
```

Would be creating a non-v201710 report definition?

Thank you.

**Comment by jradcliff** [comment]
Hi,

Could you share the code you're using to get an instance of `ReportDownloader` (or `ReportDownloaderInterface`? I just checked out the `3.10.0` tag and ran the `DownloadCriteriaReportWithSelector.java` example successfully, so I'm curious to see how your code differs from that example.

Thanks,
Josh, AdWords API Team

**Comment by Kashinoga** [comment]
Hello,

This is what our call looks like:
```
    ReportDownloadResponse response =
        new ReportDownloader(session).downloadReport(reportDefinition);
```

The downloadReport method comes from package com.google.api.ads.adwords.lib.utils.v201710. If this was not what you meant and are looking for something more detailed, please let me know.

**Comment by jradcliff** [comment]
Hi,

I just reached out to you for more information via the private discussion from the forum thread. Let's continue the discussion there. I'll come back and update this issue once we have a solution.

Thanks,
Josh, AdWords API Team

**Comment by jradcliff** [comment]
Hi,

The fix for this issue was included in [3.14.0](https://github.com/googleads/googleads-java-lib/releases/tag/3.14.0). Please use that version and reopen this issue if you still encounter problems.

Thanks,
Josh, AdWords API Team
